File: modules/routing-patterns/coordinator-routing.workflow.py
# Copyright (c) 2025 Marco Fago
# https://www.linkedin.com/in/marco-fago/
#
# This code is licensed under the MIT License.
# See the LICENSE file in the repository for the full license text.

from __future__ import annotations

from typing import Any

from langchain_core.output_parsers import StrOutputParser
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.runnables import RunnableBranch, RunnablePassthrough
from langchain_google_genai import ChatGoogleGenerativeAI


def create_llm() -> ChatGoogleGenerativeAI | None:
    """Initialize the Google GenAI chat model or return None on failure."""
    try:
        llm: ChatGoogleGenerativeAI = ChatGoogleGenerativeAI(
            model="gemini-2.5-flash",
            temperature=0,
        )
        print(f"Language model initialized: {llm.model}")
        return llm
    except Exception as error:  # noqa: BLE001
        print(f"Error initializing language model: {error}")
        return None


def booking_handler(request: str) -> str:
    """Simulates the Booking Agent handling a request."""
    print("\n--- DELEGATING TO BOOKING HANDLER ---")
    return (
        f"Booking Handler processed request: '{request}'. "
        "Result: Simulated booking action."
    )


def info_handler(request: str) -> str:
    """Simulates the Info Agent handling a request."""
    print("\n--- DELEGATING TO INFO HANDLER ---")
    return (
        f"Info Handler processed request: '{request}'. "
        "Result: Simulated information retrieval."
    )


def unclear_handler(request: str) -> str:
    """Handles requests that couldn't be delegated."""
    print("\n--- HANDLING UNCLEAR REQUEST ---")
    return f"Coordinator could not delegate request: '{request}'. Please clarify."


def extract_request(payload: dict[str, Any]) -> str:
    """Extract the original user request from delegation payload."""
    request_block: dict[str, Any] = payload["request"]
    request_value: str = str(request_block["request"])
    return request_value


def decision_matches(payload: dict[str, Any], expected: str) -> bool:
    """Check whether router decision matches expected branch id."""
    decision: str = str(payload["decision"]).strip()
    return decision == expected


llm: ChatGoogleGenerativeAI | None = create_llm()

coordinator_router_prompt: ChatPromptTemplate = ChatPromptTemplate.from_messages(
    [
        (
            "system",
            """Analyze the user's request and determine which specialist handler should process it.
- If the request is related to booking flights or hotels, output 'booker'.
- For all other general information questions, output 'info'.
- If the request is unclear or doesn't fit either category, output 'unclear'.
ONLY output one word: 'booker', 'info', or 'unclear'.""",
        ),
        ("user", "{request}"),
    ]
)

coordinator_router_chain: Any | None = None
delegation_branch: RunnableBranch | None = None
coordinator_agent: Any | None = None

if llm is not None:
    coordinator_router_chain = coordinator_router_prompt | llm | StrOutputParser()

    branches: dict[str, Any] = {
        "booker": RunnablePassthrough.assign(
            output=lambda payload: booking_handler(extract_request(payload))
        ),
        "info": RunnablePassthrough.assign(
            output=lambda payload: info_handler(extract_request(payload))
        ),
        "unclear": RunnablePassthrough.assign(
            output=lambda payload: unclear_handler(extract_request(payload))
        ),
    }

    delegation_branch = RunnableBranch(
        (
            lambda payload: decision_matches(payload, "booker"),
            branches["booker"],
        ),
        (
            lambda payload: decision_matches(payload, "info"),
            branches["info"],
        ),
        branches["unclear"],
    )

    coordinator_agent = {
        "decision": coordinator_router_chain,
        "request": RunnablePassthrough(),
    } | delegation_branch | (lambda payload: payload["output"])


def main() -> None:
    """Run example requests through the coordinator agent."""
    if llm is None or coordinator_agent is None:
        print("\nSkipping execution due to LLM initialization failure.")
        return

    print("--- Running with a booking request ---")
    request_a: str = "Book me a flight to London."
    result_a: str = coordinator_agent.invoke({"request": request_a})
    print(f"Final Result A: {result_a}")

    print("\n--- Running with an info request ---")
    request_b: str = "What is the capital of Italy?"
    result_b: str = coordinator_agent.invoke({"request": request_b})
    print(f"Final Result B: {result_b}")

    print("\n--- Running with an unclear request ---")
    request_c: str = "Tell me about quantum physics."
    result_c: str = coordinator_agent.invoke({"request": request_c})
    print(f"Final Result C: {result_c}")


if __name__ == "__main__":
    main()

File: modules/02-routing-patterns/coordinator-routing.workflow.ts
/**
 * Run:
 * cd modules && node --env-file=.env --experimental-strip-types ./routing-patterns/coordinator-routing.workflow.example.ts
 */
import { StringOutputParser } from "@langchain/core/output_parsers";
import { ChatPromptTemplate } from "@langchain/core/prompts";
import {
  RunnableBranch,
  RunnablePassthrough,
  RunnableSequence,
  type RunnableAssign,
} from "@langchain/core/runnables";
import { ChatOpenAI } from "@langchain/openai";
import { resolve } from "node:path";
import { pathToFileURL } from "node:url";

const ROUTER_SYSTEM_PROMPT: string = [
  "Analyze the user's request and determine which specialist handler should process it.",
  "- If the request is related to booking flights or hotels, output 'booker'.",
  "- For all other general information questions, output 'info'.",
  "- If the request is unclear or doesn't fit either category, output 'unclear'.",
  "ONLY output one word: 'booker', 'info', or 'unclear'.",
].join("\n");

const DEFAULT_WORKFLOW_REQUESTS: Array<string> = [
  "Book me a flight to London.",
  "Book a hotel in Busan for next weekend.",
  "What is the capital of Italy?",
  "Please handle this.",
];

type CoordinatorDecision = "booker" | "info" | "unclear";
type CoordinatorAgentName = "booking_agent" | "info_agent" | "unclear_agent";

export type RuntimeConfig = {
  apiKey: string;
  baseUrl: string;
  modelName: string;
};

export type CoordinatorRoutingResult = {
  request: string;
  decision: CoordinatorDecision;
  output: string;
  agentName: CoordinatorAgentName;
  modelName: string;
};

type CoordinatorRequest = {
  request: string;
};

type CoordinatorBranchPayload = {
  decision: string;
  request: CoordinatorRequest;
};

type CoordinatorBranchResult = CoordinatorBranchPayload & {
  output: string;
  agentName: CoordinatorAgentName;
};

type CoordinatorAgent = {
  invoke(input: CoordinatorRequest): Promise<CoordinatorBranchResult>;
};

type RouterDecisionChain = {
  invoke(input: CoordinatorRequest): Promise<string>;
};

type DelegationAssignment = RunnableAssign<
  CoordinatorBranchPayload,
  CoordinatorBranchResult
>;

type CoordinatorRequestHandler = (requestText: string) => string;

function requireEnvValue(env: NodeJS.ProcessEnv, key: string): string {
  const envValue: string | undefined = env[key]?.trim();
  if (!envValue) {
    throw new Error(`${key} is missing.`);
  }
  return envValue;
}

function buildRuntimeConfigFromEnv(env: NodeJS.ProcessEnv): RuntimeConfig {
  return {
    apiKey: requireEnvValue(env, "API_KEY"),
    baseUrl: requireEnvValue(env, "BASE_URL"),
    modelName: requireEnvValue(env, "MODEL"),
  };
}

function bookingHandler(requestText: string): string {
  return `Booking Handler processed request: '${requestText}'. Result: Simulated booking action.`;
}

function infoHandler(requestText: string): string {
  return `Info Handler processed request: '${requestText}'. Result: Simulated information retrieval.`;
}

function unclearHandler(requestText: string): string {
  return `Coordinator could not delegate request: '${requestText}'. Please clarify.`;
}

const HANDLER_BY_AGENT_TYPE: Record<
  CoordinatorAgentName,
  CoordinatorRequestHandler
> = {
  booking_agent: bookingHandler,
  info_agent: infoHandler,
  unclear_agent: unclearHandler,
};

function getHandlerByAgentType(
  agentType: CoordinatorAgentName,
): CoordinatorRequestHandler {
  return HANDLER_BY_AGENT_TYPE[agentType];
}

function normalizeDecision(rawDecision: string): CoordinatorDecision {
  const normalizedDecision: string = rawDecision.trim().toLowerCase();
  if (normalizedDecision === "booker") {
    return "booker";
  }
  if (normalizedDecision === "info") {
    return "info";
  }
  return "unclear";
}

function createDelegationAssignment(
  agentName: CoordinatorAgentName,
): DelegationAssignment {
  const handler: CoordinatorRequestHandler = getHandlerByAgentType(agentName);
  return RunnablePassthrough.assign({
    output: (input: CoordinatorBranchPayload): string =>
      handler(input.request.request),
    agentName: (_input: CoordinatorBranchPayload): CoordinatorAgentName =>
      agentName,
  });
}

function createRouterPrompt(): ReturnType<
  typeof ChatPromptTemplate.fromMessages
> {
  return ChatPromptTemplate.fromMessages([
    ["system", ROUTER_SYSTEM_PROMPT],
    ["user", "{request}"],
  ]);
}

function createRouterModel(runtimeConfig: RuntimeConfig): ChatOpenAI {
  return new ChatOpenAI({
    apiKey: runtimeConfig.apiKey,
    model: runtimeConfig.modelName,
    temperature: 0,
    timeout: 180000,
    maxRetries: 1,
    configuration: {
      baseURL: runtimeConfig.baseUrl,
    },
  });
}

function createRouterParser(): StringOutputParser {
  return new StringOutputParser();
}

function createRouterDecisionChain(
  runtimeConfig: RuntimeConfig,
): RouterDecisionChain {
  const prompt: ReturnType<typeof ChatPromptTemplate.fromMessages> =
    createRouterPrompt();
  const model: ChatOpenAI = createRouterModel(runtimeConfig);
  const parser: StringOutputParser = createRouterParser();
  return prompt.pipe(model).pipe(parser) as RouterDecisionChain;
}

function createCoordinatorDelegationBranch(): RunnableBranch<
  CoordinatorBranchPayload,
  CoordinatorBranchResult
> {
  return RunnableBranch.from<CoordinatorBranchPayload, CoordinatorBranchResult>(
    [
      [
        (input: CoordinatorBranchPayload): boolean =>
          normalizeDecision(input.decision) === "booker",
        createDelegationAssignment("booking_agent"),
      ],
      [
        (input: CoordinatorBranchPayload): boolean =>
          normalizeDecision(input.decision) === "info",
        createDelegationAssignment("info_agent"),
      ],
      createDelegationAssignment("unclear_agent"),
    ],
  );
}

function createCoordinatorSequence(
  routerChain: RouterDecisionChain,
  delegationBranch: RunnableBranch<
    CoordinatorBranchPayload,
    CoordinatorBranchResult
  >,
): CoordinatorAgent {
  return RunnableSequence.from([
    {
      decision: routerChain,
      request: new RunnablePassthrough<CoordinatorRequest>(),
    },
    delegationBranch,
  ]) as CoordinatorAgent;
}

function buildCoordinatorAgent(runtimeConfig: RuntimeConfig): CoordinatorAgent {
  const routerChain: RouterDecisionChain =
    createRouterDecisionChain(runtimeConfig);
  const delegationBranch: RunnableBranch<
    CoordinatorBranchPayload,
    CoordinatorBranchResult
  > = createCoordinatorDelegationBranch();
  return createCoordinatorSequence(routerChain, delegationBranch);
}

function runCoordinatorAgentForRequest(
  coordinatorAgent: CoordinatorAgent,
  requestText: string,
): Promise<CoordinatorBranchResult> {
  return coordinatorAgent.invoke({
    request: requestText,
  });
}

function toRoutingResult(
  requestText: string,
  modelName: string,
  branchResult: CoordinatorBranchResult,
): CoordinatorRoutingResult {
  return {
    request: requestText,
    decision: normalizeDecision(branchResult.decision),
    output: branchResult.output,
    agentName: branchResult.agentName,
    modelName,
  };
}

export async function runCoordinatorWorkflow(
  env: NodeJS.ProcessEnv = process.env,
): Promise<Array<CoordinatorRoutingResult>> {
  const runtimeConfig: RuntimeConfig = buildRuntimeConfigFromEnv(env);

  const coordinatorAgent: CoordinatorAgent =
    buildCoordinatorAgent(runtimeConfig);

  const results: Array<CoordinatorRoutingResult> = [];

  for (const requestText of DEFAULT_WORKFLOW_REQUESTS) {
    const branchResult: CoordinatorBranchResult =
      await runCoordinatorAgentForRequest(coordinatorAgent, requestText);
    const routingResult: CoordinatorRoutingResult = toRoutingResult(
      requestText,
      runtimeConfig.modelName,
      branchResult,
    );
    results.push(routingResult);
  }

  return results;
}

function isDirectExecution(
  moduleUrl: string,
  argv: readonly string[],
): boolean {
  const scriptPath: string | undefined = argv[1];
  if (!scriptPath) {
    return false;
  }

  const scriptUrl: string = pathToFileURL(resolve(scriptPath)).href;
  return scriptUrl === moduleUrl;
}

async function runCoordinatorWorkflowFromCli(): Promise<void> {
  const results: Array<CoordinatorRoutingResult> = await runCoordinatorWorkflow(
    process.env,
  );
  console.log(JSON.stringify(results, null, 2));
}

async function runCoordinatorWorkflowFromCliIfDirectExecution(
  moduleUrl: string,
  argv: readonly string[],
): Promise<void> {
  const isDirectMode: boolean = isDirectExecution(moduleUrl, argv);
  if (!isDirectMode) {
    return;
  }

  await runCoordinatorWorkflowFromCli();
}

void runCoordinatorWorkflowFromCliIfDirectExecution(
  import.meta.url,
  process.argv,
).catch((error: unknown): void => {
  const errorMessage: string =
    error instanceof Error ? error.message : String(error);
  console.error(errorMessage);
  process.exitCode = 1;
});
